src/components/ActionMenu/components/Actions/Actions.tsx

```diff
@@ -35,32 +56,92 @@ export function Actions({actions = [], groups = []}: Props) {
     () => setActiveMenuGroup(undefined),
     [],
   );
-  const menuActions = [...actions, ...groups];
 
+  // eslint-disable-next-line react-hooks/exhaustive-deps
+  const handleResize = useCallback(
+    debounce(
+      () => {
+        if (!newDesignLanguage || !actionsLayoutRef.current) return;
+        setAvailableWidth(actionsLayoutRef.current.offsetWidth);
+      },
+      20,
+      {leading: false, trailing: true, maxWait: 40},
+    ),
+    [actionsLayoutRef],
+  );
+
+  useEffect(() => {
+    if (!actionsLayoutRef.current) return;
+    setAvailableWidth(actionsLayoutRef.current.offsetWidth);
+  }, [actionsLayoutRef]);
+
+  useEffect(() => {
+    if (!menuGroupRef.current) return;
+    menuGroupWidthRef.current = menuGroupRef.current?.offsetWidth;
+  }, [menuGroupRef]);
+
+  useEffect(() => {
+    if (!newDesignLanguage || actionWidths.length === 0 || availableWidth === 0)
+      return;
+
+    let currentAvailableWidth = availableWidth;
+    setShowableActions([]);
+    setHiddenActions([]);
+
+    actions.forEach((action, index) => {
+      if (
+        actionWidths[index] + menuGroupWidthRef.current + ACTION_SPACING <=
+        currentAvailableWidth
+      ) {
+        currentAvailableWidth -= actionWidths[index];
+        setShowableActions((showableActions) => [
+          ...(showableActions || []),
+          action,
+        ]);
+      } else {
+        currentAvailableWidth = 0;
+        setHiddenActions((hiddenActions) => [...hiddenActions, action]);
+      }
+    });
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [availableWidth]);
+
+  const className = classNames(
+    styles.ActionsLayout,
+    newDesignLanguage && styles.newDesignLanguage,
+  );
+
+  const menuActions = [...actions, ...groups];
   const overriddenActions = sortAndOverrideActionOrder(menuActions);
```

Comment:
This is a good candidate for a refactor but I don't have all of the context on it. In the future I'd like to remove this function, and render actions and groups separately. As far as I can tell this function merges the actions and groups and make sure the items are rendered before the groups in proper sequence. It seems like a lot of added complexity for not much gain

Reply:
Want to make an issue for this and put it into fast follows/v6? Would love to track this tech debt.

Reply:
Added it as part of https://github.com/Shopify/polaris-ux/issues/470